### ISSUE TITLE
Render user-friendly string representation of open files

### ISSUE THREAD

**Issue by piskvorky** [RaRe-Technologies/smart_open#316]
Currently, an open S3 object prints like this:

```python
>>> f = smart_open.open('s3://…')
>>> print(f)
<smart_open.s3.SeekableBufferedInputBase object at 0x10ad02b38>
```

This is not useful, and we can do much better: [we have full info](https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/s3.py#L329) about the concrete bucket, key, mode, etc. Compare to boto3's nice native display:

```python
>>> print(f._object)
s3.Object(bucket_name='some-bucket', key='test/bank_form.pdf')
```

Task: render open objects more nicely in `__str__`. This is **great help during logging and debugging** (I consider a human-friendly `str()` best practice, and define them religiously on custom classes).

It seems easy enough to do for S3, but I'm not sure how much work for our other supported storages (HTTP, FTP or whatever), for consistency.

**Comment by mpenkov** [comment]
We could implement `__str__` and `__repr__` for our custom classes, which would solve the problem partially.

What should we do for non-custom wrapper classes? e.g.

```python
import sys
from smart_open import open

print('str read_text: ' + str(open(sys.argv[1])))
```

gives

```
str read_text: <encodings.utf_8.StreamReader object at 0x7fa0b1c14eb8>
```

We don't really have control of that object. What's the best thing to do there?

**Comment by piskvorky** [comment]
I'd leave it as it is. "Falling back to built-in where possible" is a well-communicated property of smart_open.

**Comment by ZlatSic** [comment]
Hi!

I'm new to OSS and am looking to contribute somewhere, mind if I give this a shot?

Just to clarify, the behavior you're looking for is to related to the `__str__` and `__repr__` methods of objects created with the `s3://...` descriptor? Which properties would you like to be shown in the representations? Just the `bucket` and `key` or additionally, some of the other, optional properties?

**Comment by mpenkov** [comment]
Closed via #359
